### PR TITLE
Alter Topic Operator to respect delete.topic.enable=false broker config

### PR DESCRIPTION
**Pull request by tomncooper**
### Type of change

- Bugfix

### Description

* Added exception recovery for disabled topic deletion exception
* Added unit tests for topic deletion disabled exception

### Checklist

- [x] Update/write design documentation in `./design`
- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by tomncooper** [comment]
AFAIK this change will not require a change to the docs, as with this change the docs statement (in the "Deleting a topic" section of the "Using the topic operator" section) that: 

> The delete.topic.enable property must be set to true in Kafka.spec.kafka.config. Otherwise, the steps outlined here will delete the KafkaTopic resource, but the Kafka topic and its data will remain. After reconciliation by the Topic Operator, the custom resource is then recreated. 

Is now actually correct.

File: topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java [hunk 1]
```diff
@@ -555,6 +556,27 @@ public TopicOperator(Vertx vertx, Kafka kafka,
...
+
+        if (thrown instanceof org.apache.kafka.common.errors.TopicDeletionDisabledException) {
+            LOGGER.warn("Topic deletion is disabled. Kafka topic will persist and K8s resource will be recreated in next reconciliation.");
+        } else {
+            LOGGER.error("Topic deletion failed with (" + thrown.getClass() + ") error: " + thrown.getMessage());
+            return Future.failedFuture(thrown);
+        }
+
+        return Future.succeededFuture();
+
```
Comment (ppatierno) [inline topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java]: blank line

File: topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java [hunk 2]
```diff
@@ -555,6 +556,26 @@ public TopicOperator(Vertx vertx, Kafka kafka,
...
+     * is set to false on the broker the exception is ignored an a blank future returned. For any other form of exception
+     * a failed future is returned using that exception as the cause.
+     *
+     * @param thrown The exception encountered when attempting to delete the kafka topic.
+     * @return Either an succeeded future in the case that topic deletion is disabled or a failed future in all other cases.
+     */
+    private Future<Void> handleTopicDeletionDisabled(Throwable thrown) {
+
+        if (thrown instanceof org.apache.kafka.common.errors.TopicDeletionDisabledException) {
+            LOGGER.warn("Topic deletion is disabled. Kafka broker topic will persist and KafkaTopic resource will be recreated in the next reconciliation.");
```
Comment (tombentley) [inline topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java]:
```suggestion
            LOGGER.warn("Topic deletion is disabled. Kafka topic will persist and KafkaTopic resource will be recreated in the next reconciliation.");
```

**Comment by tomncooper** [comment]
Whoops, I mistakenly rebased this branch on upstream master and pushed it. This looks like a bit of a mess now. Also because of the DCO accepting suggested edits is not a good idea! Should I blow this PR away and start again or is this fixable with git foo?

**Comment by tomncooper** [comment]
I have added a new integration test class for the TO when topic deletion is disabled. It extends the base TO IT, as a lot of the setup is shared, however it may be better to create a shared class that both TO ITs use for setup?

**Comment by scholzj** [comment]
@tombentley Does this look ok now?

File: topic-operator/src/test/java/io/strimzi/operator/topic/TopicOperatorTopicDeletionDisabledIT.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+@RunWith(VertxUnitRunner.class)
+public class TopicOperatorTopicDeletionDisabledIT extends TopicOperatorBaseIT {
+
+    private static final Logger LOGGER = LogManager.getLogger(TopicOperatorTopicDeletionDisabledIT.class);
+
+    @Before
+    public void setup(TestContext context) throws Exception {
```
Comment (tombentley) [inline topic-operator/src/test/java/io/strimzi/operator/topic/TopicOperatorTopicDeletionDisabledIT.java]: I suspect we could extract a method for starting the kafka cluster from this which we put in the base class, but I'm happy for that to be done in an other PR if the need to merge this is pressing.

**Comment by scholzj** [comment]
The DCO doesn't like the sign off on the suggested change commit. Not sure who should actually sign it in this case. Any ideas what the best approach is @tombentley?

**Comment by tombentley** [comment]
I think this should fall under clause (b) of https://developercertificate.org/: @tomncooper has signed his commits, I think we can take it on trust that any of the _maintainers_ would sign their suggestions if github had a mechanism for doing that, so overall I think it would be OK, because every commit would be certified by someone. What I'm not completely clear on is how that ought to work when squashing commits when merging (do we signoff for Tom's commits? Or should we have two signed-off-by lines?). I guess so long as that commit it signed off by someone it doesn't really matter. Thus it would be OK for you to signoff on the squashed commit.

**Comment by scholzj** [comment]
Ok, in that case I would suggest to merge it. Should we have an issue to your last comment?

**Comment by tombentley** [comment]
I can address it in my PR for TO status, but I'll open an issue in any case.

**Pull request merged**